### DISCUSSION TITLE
Atlas Scientific EZO modules, multiple ORP modules

### DISCUSSION THREAD

**Discussion by SimonFili** [arendst/Tasmota#18746] [Support request]
Hi all,

I properly compiled Tasmota32 S3 and enabled the EZO ORP module.  I can see it
![image](https://github.com/arendst/Tasmota/assets/7669537/f1f3954d-4169-41ac-af59-637018cc28bb)

My issue is that I need to read multiple EZO ORP module that have multiple I2C address. 

Right now, I can only see one.

15:50:27.696 CMD: i2cscan
15:50:28.702 MQT: stat/ESP32S3_T1/RESULT = {"I2CScan":"Error 5 at 0x02"}

I know we can run multiple EZO modules at different I2C address. I did before pH, DO, EC.

What would be needed in the code of these modules to enable reading many of the same module on different address?

Thanks!

**Comment by sfromis** [comment]
If I2CScan fails, that is usually due to "electrical" issues, like problematic wiring.

**Reply by SimonFili** [reply]
With an old version of Tasmota, I could read the 3 modules address.
With a generic latest esp32s3 version, I get this error: {"I2CScan":"Error 5 at 0x02"}
With a compiled version with EZO ORP enabled, I also get the error, but I can read one of the 3 modules installed.

If I physically remove 2 modules and keep the one that worked: I still get the same error and I can properly read the mV

I then tested the other modules, one by one:
1) {"I2CScan":"Error 5 at 0x02"}  -- I can read the mV  ( only one that flashgreen/blue)
2) {"I2CScan":"Error 5 at 0x61"}  -- no reading
3) {"I2CScan":"Error 5 at 0x62"}  -- no reading.

Hum... not sure what's happening.

**Comment by SimonFili** [comment]
@sfromis I made some progress, but still have an error on i2cScan

I connected on EZO at a time, configured separate address, now I can see all EZO.  I recompiled to enable all those EZO.

00:00:03.094 I2C: EZOEC found at 0x61
00:00:03.099 I2C: EZOORP found at 0x62
00:00:03.104 I2C: EZOORP found at 0x63
00:00:03.109 I2C: EZOORP found at 0x64
00:00:03.114 I2C: EZOORP found at 0x65
00:00:03.119 I2C: EZOpH found at 0x66

07:36:51.778 CMD: i2cscan
07:36:52.783 MQT: stat/ESP32S3_T1/RESULT = {"I2CScan":"Error 5 at 0x02"}

I disabled all other drivers, just kept 55 (EZO)
07:38:39.101 CMD: i2cdriver
07:38:39.105 MQT: stat/ESP32S3_T1/RESULT = {"I2CDriver":"!7,!8,!9,!10,!11,!12,!13,!14,!15,!17,!18,!20,!24,!29,!31,!36,!41,!42,!44,!46,!48,55,!58,!62,!65,!69,!76,!77"}

I'm sure it's not an electrical issue, since I can read everything and i'm using a board to connect the EZO.
![image](https://github.com/arendst/Tasmota/assets/7669537/37073058-a20b-4c3f-8f23-47890c2d4c2e)

Any other ideas?

**Reply by sfromis** [reply]
While I have no experience with EZO sensors, or looked into how the combined driver setup works, I observe that the [EZOORP driver](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xsns_sensor/xsns_78_ezoorp.ino) does not have support for a per-sensor id, which is usually when a driver just does not support more than one of the same type. Just a guess.....

**Reply by SimonFili** [reply]
Thanks for the info.  I'm unsure I understand everything around the drivers and per-sensor id's but here's what I see now:
![image](https://github.com/arendst/Tasmota/assets/7669537/0963e59a-96f4-46d6-b288-60f63b723bd2)

Therefore, I see all EZO ORP individually.  

I see this pushed to MQTT
{
    "ANALOG": {
        "A1": 4095
    },
    "EZO-1": {
        "EC": 0.0
    },
    "EZO-2": {
        "ORP": 74.7
    },
    "EZO-3": {
        "ORP": 1020.0
    },
    "EZO-4": {
        "ORP": 599.5
    },
    "EZO-5": {
        "ORP": 1020.0
    },
    "EZO-6": {
        "pH": 0.0
    },
    "MHZ19B": {
        "CarbonDioxide": 400,
        "Model": "B",
        "Temperature": 25.0
    },
    "TempUnit": "C",
    "Time": "2023-05-30T09:45:52"
}

The issue I have now that there's no unit for the EZO ORP, even though the web display show mV.

I need to find a way to add the unit on the MQTT message.

Thanks for participating!

**Reply by sfromis** [reply]
The JSON only includes a unit for the very few values where you have an option to chose which you want (temperature and atmospheric pressure). In other cases, the subscriber is assumed to already know what to expect.

**Reply by SimonFili** [reply]
Ok thanks, I'm taking care of that on the MQTT receiving side, Home Assistant in that case.
